addon/components/base/bs-dropdown.js

```diff
@@ -272,61 +283,88 @@ export default class Dropdown extends Component {
   @action
   toggleDropdown() {
     if (this.get('isOpen')) {
-      this.send('closeDropdown');
+      this.closeDropdown();
     } else {
-      this.send('openDropdown');
+      this.openDropdown();
     }
   }
 
   @action
   openDropdown() {
     this.set('isOpen', true);
-    this.addClickListener();
     this.get('onShow')();
   }
 
   @action
   closeDropdown() {
     this.set('isOpen', false);
-    this.removeClickListener();
     this.get('onHide')();
   }
 
-  addClickListener() {
-    if (!this.clickListener) {
-      this.clickListener = bind(this, this.closeOnClickHandler);
-      document.addEventListener('click', this.clickListener, true);
-    }
-  }
-
-  removeClickListener() {
-    if (this.clickListener)  {
-      document.removeEventListener('click', this.clickListener, true);
-      this.clickListener = null;
-    }
-  }
-
-  willDestroyElement() {
-    super.willDestroyElement(...arguments);
-    this.removeClickListener();
-  }
-
   /**
    * Handler for click events to close the dropdown
    *
    * @method closeOnClickHandler
    * @param e
    * @protected
    */
-  closeOnClickHandler(e) {
+  @action
+  closeHandler(e) {
     let { target } = e;
     let { toggleElement, menuElement } = this.getProperties('toggleElement', 'menuElement');
 
     if (!this.get('isDestroyed')
-      && (toggleElement && !toggleElement.contains(target))
-      && ((menuElement && !menuElement.contains(target)) || this.get('closeOnMenuClick'))) {
-      this.send('closeDropdown');
+      && (
+        (e.type === 'keyup' && e.which === TAB_KEYCODE && menuElement && !menuElement.contains(target))
+        || (e.type === 'click' && toggleElement && !toggleElement.contains(target) && ((menuElement && !menuElement.contains(target)) || this.get('closeOnMenuClick')))
+      )) {
+      this.closeDropdown();
+    }
+  }
+
+  @action
+  handleKeyEvent(event) {
+    if (/input|textarea/i.test(event.target.tagName)
```

Comment:
IMO `Array.includes()` is easier to read but that's just coding style.

```suggestion
    if (['input', 'textarea'].includes(event.target.tagName)
```

Reply:
Yeah, I copy pasted this from bootstrap.js tbh 😆

Reply:
Sorry, my bad. The regex was case insensitive. `tagName` is uppercase.

```suggestion
    if (/input|textarea/i.test(event.target.tagName.toLowerCase())
```

Reply:
Fixed.